File: 과제10_detroit/detroit.py
def inputData():
    inp = open("detroit.inp")
    N = int(inp.readline())
    arr = [[[] for _ in range(N)] for _ in range(N)]
    for i in range(N):
        temp = list(map(int, "".join(inp.readline().split())))
        for j in range(N):
            if temp[j] == 0:
                arr[i][j] = []
            else:
                arr[i][j] = [temp[j]]
    inp.close()
    for i in range(N):
        for j in range(N):
            if not arr[i][j]:
                temp = [i + 1 for i in range(N)]
                for k in range(N):
                    if arr[i][k] and len(arr[i][k]) == 1:
                        if arr[i][k][0] in temp:
                            temp.remove(arr[i][k][0])
                    if arr[k][j] and len(arr[k][j]) == 1:
                        if arr[k][j][0] in temp:
                            temp.remove(arr[k][j][0])
                arr[i][j] = temp
    return arr


def outputData(answer):
    out = open("detroit.out", "w")
    out.write("%d" % answer)
    out.close()
    return


def put_topping(pizza, number, a, b):
    if b == len(pizza):
        a += 1
        b = 0
    if a == len(pizza) and b == 0:
        return number + 1
    if len(pizza[a][b]) == 1:
        number = put_topping(pizza, number, a, b + 1)

    temp = pizza[a][b].copy()
    temp2 = temp.copy()
    while 1:
        if not temp:
            pizza[a][b] = temp2
            return number
        for i in range(len(pizza)):
            if len(pizza[a][i]) == 1:
                if pizza[a][i][0] in temp:
                    temp.remove(pizza[a][i][0])
            if len(pizza[i][b]) == 1:
                if pizza[i][b][0] in temp:
                    temp.remove(pizza[i][b][0])
        for i in range(len(temp)):
            pizza[a][b] = [temp[i]]
            number = put_topping(pizza, number, a, b + 1)
        temp = []


if __name__ == "__main__":
    outputData(put_topping(inputData(), 0, 0, 0))

File: 과제10_detroit/detroit.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <cstdio>
#include <vector>
#include <iostream>
#include <time.h>
#include <algorithm>

using namespace std;
struct Tup {
	int a = 0, b = 0, c = 0, temp[9] = { 1,1,1,1,1,1,1,1,1 };
	Tup* next;
};
int pizza[9][9], N, cases;
vector<Tup*> index;

bool sorting(Tup* A, Tup* B) {
	return A->c < B->c ? true : false;
}

void inputdata() {
	FILE* inp = fopen("sampleData/4.inp", "r");
	fscanf(inp, "%d", &N);
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			fscanf(inp, "%d", &pizza[i][j]);
		}
	}
	fclose(inp);
	return;
}
void outputdata(int a) {
	FILE* out = fopen("detroit.out", "w");
	if (a == -1) {
		fprintf(out, "%d", 0);
		printf("%d\n", 0);
	}
	else if (a == 1) {
		fprintf(out, "%d", 1);
		printf("%d\n", 1);
	}
	else {
		fprintf(out, "%d", cases);
		printf("%d\n", cases);
	}
	fclose(out);
}

int makefirstpizza() {
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			if (pizza[i][j] == 0) {
				Tup* t = new Tup;
				t->a = i;
				t->b = j;
				for (int k = 0; k < N; k++) {
					if (pizza[i][k] != 0) {
						t->temp[pizza[i][k] - 1] = 0;
					}
					if (pizza[k][j] != 0) {
						t->temp[pizza[k][j] - 1] = 0;
					}
				}
				for (int k = 0; k < N; k++) {
					if (t->temp[k] == 1) {
						t->c++;
					}
				}
				if (t->c == 0) {
					return -1;
				}
				index.push_back(t);
			}
			else if (pizza[i][j] != 0) {
				for (int k = 0; k < N; k++) {
					if (pizza[i][j] != 0 && j != k && pizza[i][j] == pizza[i][k]) {
						return -1;
					}
					if (pizza[i][j] != 0 && i != k && pizza[i][j] == pizza[k][j]) {
						return -1;
					}
				}
			}
		}
	}
	if (index.size() == 0) {
		return 1;
	}
	sort(index.begin(), index.end(), sorting);
	for (int i = 0; i < index.size(); i++) {
		if (i == index.size() - 1) {
			index[i]->next = nullptr;
		}
		else {
			index[i]->next = index[i + 1];
		}
	}
	return 0;
}

void put_topping(Tup* key) {
	if (key->next == nullptr) {
		cases++;
		return;
	}
	int a = key->a, b = key->b;
	for (int i = 0; i < N; i++) {
		if (pizza[a][i] != 0) {
			key->temp[pizza[a][i] - 1] = 0;
		}
		if (pizza[i][b] != 0) {
			key->temp[pizza[i][b] - 1] = 0;
		}
	}
	for (int i = 0; i < N; i++) {
		if (key->temp[i] == 1) {
			pizza[a][b] = i + 1;
			put_topping(key->next);
			pizza[a][b] = 0;
		}
	}
	return;
}

int main() {
	clock_t start, end;
	start = clock();
	inputdata();
	int flag = makefirstpizza();
	if (flag == -1) {
		outputdata(-1);
	}
	else if (flag == 1) {
		outputdata(1);
	}
	else {
		put_topping(index[0]);
		outputdata(0);
	}
	end = clock();
	cout << (double)(end - start) / CLOCKS_PER_SEC;
}